File: clients/java/src/test/java/io/volant/DescribeAlterConfigsTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** DescribeConfigs / AlterConfigs client tests against a scripted TCP broker. */
class DescribeAlterConfigsTest {
    @Test
    void describeReturnsPairs() throws Exception {
        try (ConfigsServer srv = ConfigsServer.ok()) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                DescribeConfigsResult got = c.describeConfigs("events");
                assertEquals("events", got.topic);
                assertEquals(1, got.topicId);
                assertEquals(1, got.partitionCount);
                assertEquals(1, got.configs.size());
                assertEquals("retention.ms", got.configs.get(0)[0]);
                assertEquals("86400000", got.configs.get(0)[1]);
            }
            assertEquals("events", srv.topic.get());
            assertEquals(Collections.singletonList(Codec.OP_DESCRIBE_CONFIGS), srv.opcodes);
        }
    }

    @Test
    void alterOkEmptyValueClear() throws Exception {
        try (ConfigsServer srv = ConfigsServer.ok()) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                c.alterConfigs("events", Collections.singletonList(new String[] {"retention.ms", ""}));
            }
            assertEquals("events", srv.topic.get());
            assertEquals(1, srv.alter.size());
            assertArrayEquals(new String[] {"retention.ms", ""}, srv.alter.get(0));
            assertEquals(Collections.singletonList(Codec.OP_ALTER_CONFIGS), srv.opcodes);
        }
    }

    @Test
    void describeErrorCodeRaises() throws Exception {
        try (ConfigsServer srv = ConfigsServer.error(2)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(BrokerException.class, () -> c.describeConfigs("missing"));
                assertEquals(2, ex.code);
                assertEquals("describe_configs", ex.op);
            }
        }
    }

    @Test
    void alterErrorCodeRaises() throws Exception {
        try (ConfigsServer srv = ConfigsServer.error(2)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(
                        BrokerException.class,
                        () -> c.alterConfigs(
                                "missing", Collections.singletonList(new String[] {"retention.ms", "1"})));
                assertEquals(2, ex.code);
                assertEquals("alter_configs", ex.op);
            }
        }
    }

    private static final class ConfigsServer implements AutoCloseable {
        final int port;
        final AtomicReference<String> topic = new AtomicReference<>();
        final List<Integer> opcodes = new CopyOnWriteArrayList<>();
        final List<String[]> alter = new CopyOnWriteArrayList<>();
        private final int errorCode;
        private final ServerSocket listen;
        private final Thread thread;

        static ConfigsServer ok() throws IOException {
            return new ConfigsServer(0);
        }

        static ConfigsServer error(int code) throws IOException {
            return new ConfigsServer(code);
        }

        private ConfigsServer(int errorCode) throws IOException {
            this.errorCode = errorCode;
            listen = new ServerSocket(0, 8, InetAddress.getByName("127.0.0.1"));
            listen.setSoTimeout(8_000);
            port = listen.getLocalPort();
            thread = new Thread(this::serve, "volant-configs");
            thread.setDaemon(true);
            thread.start();
        }

        private void serve() {
            try (Socket conn = listen.accept()) {
                conn.setSoTimeout(5_000);
                InputStream in = conn.getInputStream();
                OutputStream out = conn.getOutputStream();
                byte[] buf = new byte[0];
                while (true) {
                    Frame.Decode d = Frame.tryDecode(buf);
                    if (d.frame == null) {
                        byte[] tmp = new byte[4096];
                        int n = in.read(tmp);
                        if (n < 0) {
                            return;
                        }
                        byte[] next = new byte[buf.length + n];
                        System.arraycopy(buf, 0, next, 0, buf.length);
                        System.arraycopy(tmp, 0, next, buf.length, n);
                        buf = next;
                        continue;
                    }
                    buf = d.rest;
                    opcodes.add(d.frame.opcode);
                    if (d.frame.opcode == Codec.OP_DESCRIBE_CONFIGS) {
                        Codec.DescribeConfigsRequest req = Codec.decodeDescribeConfigsRequest(d.frame.payload);
                        topic.set(req.topic);
                        byte[] payload = Codec.encodeDescribeConfigsResponse(
                                new Codec.DescribeConfigsResponse(
                                        errorCode,
                                        req.topic,
                                        1,
                                        1,
                                        Collections.singletonList(new String[] {"retention.ms", "86400000"})));
                        out.write(Frame.encode(
                                Codec.OP_DESCRIBE_CONFIGS_RESPONSE, d.frame.correlationId, payload));
                        out.flush();
                    } else if (d.frame.opcode == Codec.OP_ALTER_CONFIGS) {
                        Codec.AlterConfigsRequest req = Codec.decodeAlterConfigsRequest(d.frame.payload);
                        topic.set(req.topic);
                        alter.clear();
                        alter.addAll(req.configs);
                        byte[] payload = Codec.encodeAlterConfigsResponse(
                                new Codec.AlterConfigsResponse(errorCode, req.topic));
                        out.write(Frame.encode(
                                Codec.OP_ALTER_CONFIGS_RESPONSE, d.frame.correlationId, payload));
                        out.flush();
                    } else {
                        return;
                    }
                }
            } catch (Exception ignored) {
                // connection close ends the script
            }
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
                // best-effort
            }
            try {
                thread.join(2_000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}

File: clients/go/describe_alter_configs_test.go
package volant_test

import (
	"errors"
	"net"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

type configsServerResult struct {
	opcodes []uint16
	topic   string
	alter   [][2]string
	err     error
}

func serveConfigs(t *testing.T, errorCode uint16, configs [][2]string) (addr string, got *configsServerResult, stop func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	res := &configsServerResult{}
	done := make(chan struct{})
	go func() {
		defer close(done)
		conn, err := ln.Accept()
		if err != nil {
			res.err = err
			return
		}
		defer conn.Close()
		_ = conn.SetDeadline(time.Now().Add(5 * time.Second))
		buf := make([]byte, 0, 4096)
		tmp := make([]byte, 4096)
		for {
			f, rest, err := frame.TryDecode(buf)
			if err != nil {
				res.err = err
				return
			}
			if f == nil {
				n, err := conn.Read(tmp)
				if n > 0 {
					buf = append(buf, tmp[:n]...)
				}
				if err != nil {
					res.err = err
					return
				}
				continue
			}
			buf = append([]byte(nil), rest...)
			res.opcodes = append(res.opcodes, f.Opcode)
			switch f.Opcode {
			case codec.OpDescribeConfigs:
				req, err := codec.DecodeDescribeConfigsRequest(f.Payload)
				if err != nil {
					res.err = err
					return
				}
				res.topic = req.Topic
				payload, err := codec.EncodeDescribeConfigsResponse(codec.DescribeConfigsResponse{
					ErrorCode:      errorCode,
					Topic:          req.Topic,
					TopicID:        1,
					PartitionCount: 1,
					Configs:        configs,
				})
				if err != nil {
					res.err = err
					return
				}
				raw, err := frame.Encode(codec.OpDescribeConfigsResponse, f.CorrelationID, payload)
				if err != nil {
					res.err = err
					return
				}
				if _, err := conn.Write(raw); err != nil {
					res.err = err
					return
				}
			case codec.OpAlterConfigs:
				req, err := codec.DecodeAlterConfigsRequest(f.Payload)
				if err != nil {
					res.err = err
					return
				}
				res.topic = req.Topic
				res.alter = req.Configs
				payload, err := codec.EncodeAlterConfigsResponse(codec.AlterConfigsResponse{
					ErrorCode: errorCode,
					Topic:     req.Topic,
				})
				if err != nil {
					res.err = err
					return
				}
				raw, err := frame.Encode(codec.OpAlterConfigsResponse, f.CorrelationID, payload)
				if err != nil {
					res.err = err
					return
				}
				if _, err := conn.Write(raw); err != nil {
					res.err = err
					return
				}
			default:
				res.err = &frame.ProtocolError{Msg: "unexpected opcode"}
				return
			}
		}
	}()
	return ln.Addr().String(), res, func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func TestDescribeConfigsReturnsPairs(t *testing.T) {
	pairs := [][2]string{{"retention.ms", "86400000"}}
	addr, got, stop := serveConfigs(t, 0, pairs)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.DescribeConfigs("events")
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.topic != "events" || len(got.opcodes) != 1 || got.opcodes[0] != codec.OpDescribeConfigs {
		t.Fatalf("wire request topic=%q opcodes=%v", got.topic, got.opcodes)
	}
	if out.Topic != "events" || out.TopicID != 1 || out.PartitionCount != 1 {
		t.Fatalf("parsed %+v", out)
	}
	if len(out.Configs) != 1 || out.Configs[0] != [2]string{"retention.ms", "86400000"} {
		t.Fatalf("configs %+v", out.Configs)
	}
}

func TestAlterConfigsOkEmptyValueClear(t *testing.T) {
	addr, got, stop := serveConfigs(t, 0, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	if err := c.AlterConfigs("events", [][2]string{{"retention.ms", ""}}); err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.topic != "events" || len(got.alter) != 1 || got.alter[0] != [2]string{"retention.ms", ""} {
		t.Fatalf("wire alter topic=%q configs=%v", got.topic, got.alter)
	}
}

func TestAlterConfigEncodesOnePair(t *testing.T) {
	addr, got, stop := serveConfigs(t, 0, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	if err := c.AlterConfig("events", "retention.ms", "1"); err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.topic != "events" || len(got.alter) != 1 || got.alter[0] != [2]string{"retention.ms", "1"} {
		t.Fatalf("wire alter topic=%q configs=%v", got.topic, got.alter)
	}
}

func TestDescribeConfigsNonzeroErrorRaises(t *testing.T) {
	addr, _, stop := serveConfigs(t, 2, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.DescribeConfigs("missing")
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 2 || be.Op != "describe_configs" {
		t.Fatalf("broker error %+v", be)
	}
}

func TestAlterConfigsNonzeroErrorRaises(t *testing.T) {
	addr, _, stop := serveConfigs(t, 2, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	err = c.AlterConfigs("missing", [][2]string{{"retention.ms", "1"}})
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 2 || be.Op != "alter_configs" {
		t.Fatalf("broker error %+v", be)
	}
}
